Basic experimental fog code extracted from jshexen2 (activate with
"gl_fog 1", see the cvars for other options):
- underwater fogging removed
- slightly cleaned-up according to original quake code
- there is some striped view in fogged areas ;(
- seems NOT working correctly without multitexturing
- obviously still needs fixing ;)  the only reason I
  revied this crap is that the glowed missiles look
  somewhat better in fog (personal liking), sigh...

diff -urNp uhexen2-20060112/hexen2/gl_dl_rlight.c uhexen2-20060112b/hexen2/gl_dl_rlight.c
--- uhexen2-20060112/hexen2/gl_dl_rlight.c	2005-10-27 09:47:12.000000000 +0300
+++ uhexen2-20060112b/hexen2/gl_dl_rlight.c	2006-01-12 15:59:53.000000000 +0200
@@ -114,6 +114,10 @@ void R_RenderDlights (void)
 
 	r_dlightframecount = r_framecount + 1;	// because the count hasn't
 						//  advanced yet for this frame
+
+	// disable drawing fog on lights
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -133,6 +137,9 @@ void R_RenderDlights (void)
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
 	glDepthMask_fp (1);
+	// re-enable fog
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
 }
 
 
diff -urNp uhexen2-20060112/hexen2/gl_dl_rmain.c uhexen2-20060112b/hexen2/gl_dl_rmain.c
--- uhexen2-20060112/hexen2/gl_dl_rmain.c	2005-10-25 02:01:15.000000000 +0300
+++ uhexen2-20060112b/hexen2/gl_dl_rmain.c	2006-01-12 15:59:53.000000000 +0200
@@ -67,6 +67,7 @@ void R_MarkLeaves (void);
 
 extern	cvar_t	v_gamma;
 extern	qboolean gl_dogamma;
+GLfloat	fog_colours[4];
 
 cvar_t	r_norefresh = {"r_norefresh","0"};
 cvar_t	r_drawentities = {"r_drawentities","1"};
@@ -99,6 +100,14 @@ cvar_t	gl_stencilshadow = {"gl_stencilsh
 cvar_t	gl_glows = {"gl_glows","1",true};
 cvar_t	gl_other_glows = {"gl_other_glows","0",true};
 cvar_t	gl_missile_glows = {"gl_missile_glows","1",true};
+cvar_t	gl_fogenable	= {"gl_fog", "0", true};
+cvar_t	gl_fogstart	= {"gl_fogstart", "400.0", true};
+cvar_t	gl_fogend	= {"gl_fogend", "2000.0", true};
+cvar_t	gl_fogdensity	= {"gl_fogdensity", "0.2", true};
+cvar_t	gl_fogred	= {"gl_fogred","0.5", true};
+cvar_t	gl_foggreen	= {"gl_foggreen","0.5", true};
+cvar_t	gl_fogblue	= {"gl_fogblue","0.5", true};
+cvar_t	gl_fogalpha	= {"gl_fogalpha", "0.5", true};
 
 extern	cvar_t	gl_ztrick;
 
@@ -1204,6 +1213,9 @@ void R_DrawAllGlows(void)
 	if (!r_drawentities.value)
 		return;
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -1222,6 +1234,9 @@ void R_DrawAllGlows(void)
 		}		
 	}
 
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
+
 	glDisable_fp (GL_BLEND);
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
@@ -1598,6 +1613,23 @@ void R_RenderScene (void)
 
 	R_MarkLeaves ();	// done here so we know if we're in water
 
+#if 0
+// WHY DID JSHEXEN2 PUT THIS HERE ??!!
+	if (gl_fogenable.value)
+	{
+		glFogi_fp(GL_FOG_MODE, GL_LINEAR);
+
+		fog_colours[0] = gl_fogred.value;
+		fog_colours[1] = gl_foggreen.value;
+		fog_colours[2] = gl_fogblue.value;
+
+		glFogfv_fp(GL_FOG_COLOR, fog_colours);
+		glFogf_fp(GL_FOG_START, gl_fogstart.value);
+		glFogf_fp(GL_FOG_END, gl_fogend.value);
+		glEnable_fp(GL_FOG);
+	}
+#endif
+
 	R_DrawWorld ();		// adds static entities to the list
 
 	S_ExtraUpdate ();	// don't let sound get messed up if going slow
@@ -1606,6 +1638,8 @@ void R_RenderScene (void)
 
 	R_DrawAllGlows();
 
+	//glDisable_fp(GL_FOG);	// already done in R_RenderDlights()
+
 	R_RenderDlights ();
 
 #if 0
@@ -1803,6 +1837,39 @@ void R_RenderView (void)
 
 	R_Clear ();
 
+	if (gl_fogenable.value)
+	{
+		if (gl_fogstart.value < 5)
+			Cvar_SetValue ("gl_fogstart", 5);
+		if (gl_fogstart.value > 4095)
+			Cvar_SetValue ("gl_fogstart", 4095);
+		if (gl_fogend.value < 6)
+			Cvar_SetValue ("gl_fogend", 6);
+		if (gl_fogend.value > 4096)
+			Cvar_SetValue ("gl_fogend", 4096);
+		if (gl_fogstart.value >= gl_fogend.value)
+			Cvar_SetValue ("gl_fogend", gl_fogstart.value + 1);
+
+		fog_colours[0] = (GLfloat) gl_fogred.value;
+		fog_colours[1] = (GLfloat) gl_foggreen.value;
+		fog_colours[2] = (GLfloat) gl_fogblue.value;
+		fog_colours[3] = (GLfloat) 1;//gl_fogalpha.value;
+
+		glFogi_fp (GL_FOG_MODE, GL_LINEAR);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP2);
+		glFogfv_fp (GL_FOG_COLOR, fog_colours);
+		glFogf_fp (GL_FOG_START, gl_fogstart.value);
+		glFogf_fp (GL_FOG_END, gl_fogend.value);
+		glFogf_fp (GL_FOG_DENSITY, gl_fogdensity.value);
+	//	glHint_fp (GL_FOG_HINT, GL_NICEST);
+		glEnable_fp (GL_FOG);
+	}
+	else
+	{
+		glDisable_fp (GL_FOG);
+	}
+
 	// render normal view
 	R_RenderScene ();
 
@@ -1818,6 +1885,9 @@ void R_RenderView (void)
 
 	R_DrawViewModel();
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	// render mirror view
 	R_Mirror ();
 
@@ -1827,6 +1897,8 @@ void R_RenderView (void)
 	{
 		R_PrintTimes ();
 	}
+
+	glDepthMask_fp(1);
 }
 
 
diff -urNp uhexen2-20060112/hexen2/gl_dl_rmisc.c uhexen2-20060112b/hexen2/gl_dl_rmisc.c
--- uhexen2-20060112/hexen2/gl_dl_rmisc.c	2005-10-25 23:04:17.000000000 +0300
+++ uhexen2-20060112b/hexen2/gl_dl_rmisc.c	2006-01-12 15:59:53.000000000 +0200
@@ -216,6 +216,15 @@ void R_Init (void)
 	Cvar_RegisterVariable (&gl_reporttjunctions);
 	Cvar_RegisterVariable (&gl_multitexture);
 
+	Cvar_RegisterVariable (&gl_fogenable);
+	Cvar_RegisterVariable (&gl_fogstart);
+	Cvar_RegisterVariable (&gl_fogend);
+	Cvar_RegisterVariable (&gl_fogdensity);
+	Cvar_RegisterVariable (&gl_fogalpha);
+	Cvar_RegisterVariable (&gl_fogred);
+	Cvar_RegisterVariable (&gl_fogblue);
+	Cvar_RegisterVariable (&gl_foggreen);
+
 	Cvar_RegisterVariable (&gl_glows);
 	Cvar_RegisterVariable (&gl_missile_glows);
 	Cvar_RegisterVariable (&gl_other_glows);
diff -urN uhexen2-20060116/hexen2/gl_func.h uhexen2-20060116a/hexen2/gl_func.h
--- uhexen2-20060116/hexen2/gl_func.h	2006-01-16 18:44:10.000000000 +0200
+++ uhexen2-20060116a/hexen2/gl_func.h	2006-01-16 18:53:30.000000000 +0200
@@ -47,6 +47,11 @@
 GL_FUNCTION(void, glColor3f, (GLfloat,GLfloat,GLfloat))
 GL_FUNCTION(void, glClearColor, (GLclampf,GLclampf,GLclampf,GLclampf))
 
+GL_FUNCTION(void, glFogf, (GLenum,GLfloat))
+GL_FUNCTION(void, glFogfv, (GLenum, const GLfloat *))
+GL_FUNCTION(void, glFogi, (GLenum,GLint))
+GL_FUNCTION(void, glFogiv, (GLenum, const GLint *))
+
 GL_FUNCTION(void, glAlphaFunc, (GLenum,GLclampf))
 GL_FUNCTION(void, glBlendFunc, (GLenum,GLenum))
 GL_FUNCTION(void, glShadeModel, (GLenum))
diff -urNp uhexen2-20060112/hexen2/glquake.h uhexen2-20060112b/hexen2/glquake.h
--- uhexen2-20060112/hexen2/glquake.h	2006-01-07 11:50:09.000000000 +0200
+++ uhexen2-20060112b/hexen2/glquake.h	2006-01-12 15:59:53.000000000 +0200
@@ -323,6 +323,16 @@ extern	cvar_t	gl_glows;
 extern	cvar_t	gl_other_glows;
 extern	cvar_t	gl_missile_glows;
 
+extern	cvar_t	gl_fogenable;
+extern	cvar_t	gl_fogstart;
+extern	cvar_t	gl_fogend;
+extern	cvar_t	gl_fogdensity;
+extern	cvar_t	gl_fogalpha;
+extern	cvar_t	gl_fogred;
+extern	cvar_t	gl_fogblue;
+extern	cvar_t	gl_foggreen;
+extern	cvar_t	gl_underwaterfog;
+
 extern	int	gl_lightmap_format;
 extern	int	gl_solid_format;
 extern	int	gl_alpha_format;
diff -urN uhexen2-20060116/hexenworld/Client/gl_func.h uhexen2-20060116a/hexenworld/Client/gl_func.h
--- uhexen2-20060116/hexenworld/Client/gl_func.h	2006-01-16 18:44:10.000000000 +0200
+++ uhexen2-20060116a/hexenworld/Client/gl_func.h	2006-01-16 18:54:09.000000000 +0200
@@ -47,6 +47,11 @@
 GL_FUNCTION(void, glColor3f, (GLfloat,GLfloat,GLfloat))
 GL_FUNCTION(void, glClearColor, (GLclampf,GLclampf,GLclampf,GLclampf))
 
+GL_FUNCTION(void, glFogf, (GLenum,GLfloat))
+GL_FUNCTION(void, glFogfv, (GLenum, const GLfloat *))
+GL_FUNCTION(void, glFogi, (GLenum,GLint))
+GL_FUNCTION(void, glFogiv, (GLenum, const GLint *))
+
 GL_FUNCTION(void, glAlphaFunc, (GLenum,GLclampf))
 GL_FUNCTION(void, glBlendFunc, (GLenum,GLenum))
 GL_FUNCTION(void, glShadeModel, (GLenum))
diff -urNp uhexen2-20060112/hexenworld/Client/glquake.h uhexen2-20060112b/hexenworld/Client/glquake.h
--- uhexen2-20060112/hexenworld/Client/glquake.h	2006-01-07 11:50:10.000000000 +0200
+++ uhexen2-20060112b/hexenworld/Client/glquake.h	2006-01-12 15:59:53.000000000 +0200
@@ -326,6 +326,16 @@ extern	cvar_t	gl_glows;
 extern	cvar_t	gl_other_glows;
 extern	cvar_t	gl_missile_glows;
 
+extern	cvar_t	gl_fogenable;
+extern	cvar_t	gl_fogstart;
+extern	cvar_t	gl_fogend;
+extern	cvar_t	gl_fogdensity;
+extern	cvar_t	gl_fogalpha;
+extern	cvar_t	gl_fogred;
+extern	cvar_t	gl_fogblue;
+extern	cvar_t	gl_foggreen;
+extern	cvar_t	gl_underwaterfog;
+
 extern	int	gl_lightmap_format;
 extern	int	gl_solid_format;
 extern	int	gl_alpha_format;
diff -urNp uhexen2-20060112/hexenworld/Client/gl_rlight.c uhexen2-20060112b/hexenworld/Client/gl_rlight.c
--- uhexen2-20060112/hexenworld/Client/gl_rlight.c	2005-10-27 09:47:12.000000000 +0300
+++ uhexen2-20060112b/hexenworld/Client/gl_rlight.c	2006-01-12 15:59:53.000000000 +0200
@@ -139,6 +139,10 @@ void R_RenderDlights (void)
 
 	r_dlightframecount = r_framecount + 1;	// because the count hasn't
 						//  advanced yet for this frame
+
+	// disable drawing fog on lights
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -158,6 +162,9 @@ void R_RenderDlights (void)
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
 	glDepthMask_fp (1);
+	// re-enable fog
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
 }
 
 
diff -urNp uhexen2-20060112/hexenworld/Client/gl_rmain.c uhexen2-20060112b/hexenworld/Client/gl_rmain.c
--- uhexen2-20060112/hexenworld/Client/gl_rmain.c	2005-10-25 02:01:15.000000000 +0300
+++ uhexen2-20060112b/hexenworld/Client/gl_rmain.c	2006-01-12 15:59:53.000000000 +0200
@@ -67,6 +67,7 @@ void R_MarkLeaves (void);
 
 extern	cvar_t	v_gamma;
 extern	qboolean gl_dogamma;
+GLfloat	fog_colours[4];
 
 cvar_t	r_norefresh = {"r_norefresh","0"};
 cvar_t	r_drawentities = {"r_drawentities","1"};
@@ -100,6 +101,14 @@ cvar_t	gl_stencilshadow = {"gl_stencilsh
 cvar_t	gl_glows = {"gl_glows","1",true};
 cvar_t	gl_other_glows = {"gl_other_glows","0",true};
 cvar_t	gl_missile_glows = {"gl_missile_glows","1",true};
+cvar_t	gl_fogenable	= {"gl_fog", "0", true};
+cvar_t	gl_fogstart	= {"gl_fogstart", "400.0", true};
+cvar_t	gl_fogend	= {"gl_fogend", "2000.0", true};
+cvar_t	gl_fogdensity	= {"gl_fogdensity", "0.2", true};
+cvar_t	gl_fogred	= {"gl_fogred","0.5", true};
+cvar_t	gl_foggreen	= {"gl_foggreen","0.5", true};
+cvar_t	gl_fogblue	= {"gl_fogblue","0.5", true};
+cvar_t	gl_fogalpha	= {"gl_fogalpha", "0.5", true};
 
 extern	cvar_t	gl_ztrick;
 extern	cvar_t	scr_fov;
@@ -1217,6 +1226,9 @@ void R_DrawAllGlows(void)
 	if (!r_drawentities.value)
 		return;
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	glDepthMask_fp (0);
 	glDisable_fp (GL_TEXTURE_2D);
 	glShadeModel_fp (GL_SMOOTH);
@@ -1235,6 +1247,9 @@ void R_DrawAllGlows(void)
 		}		
 	}
 
+	if (gl_fogenable.value)
+		glEnable_fp(GL_FOG);
+
 	glDisable_fp (GL_BLEND);
 	glEnable_fp (GL_TEXTURE_2D);
 	glBlendFunc_fp (GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
@@ -1672,6 +1687,23 @@ void R_RenderScene (void)
 
 	R_MarkLeaves ();	// done here so we know if we're in water
 
+#if 0
+// WHY DID JSHEXEN2 PUT THIS HERE ??!!
+	if (gl_fogenable.value)
+	{
+		glFogi_fp(GL_FOG_MODE, GL_LINEAR);
+
+		fog_colours[0] = gl_fogred.value;
+		fog_colours[1] = gl_foggreen.value;
+		fog_colours[2] = gl_fogblue.value;
+
+		glFogfv_fp(GL_FOG_COLOR, fog_colours);
+		glFogf_fp(GL_FOG_START, gl_fogstart.value);
+		glFogf_fp(GL_FOG_END, gl_fogend.value);
+		glEnable_fp(GL_FOG);
+	}
+#endif
+
 	R_DrawWorld ();		// adds static entities to the list
 
 	S_ExtraUpdate ();	// don't let sound get messed up if going slow
@@ -1680,6 +1712,8 @@ void R_RenderScene (void)
 
 	R_DrawAllGlows();
 
+	//glDisable_fp(GL_FOG);	// already done in R_RenderDlights()
+
 	R_RenderDlights ();
 
 #ifdef GLTEST
@@ -1875,6 +1909,39 @@ void R_RenderView (void)
 
 	R_Clear ();
 
+	if (gl_fogenable.value)
+	{
+		if (gl_fogstart.value < 5)
+			Cvar_SetValue ("gl_fogstart", 5);
+		if (gl_fogstart.value > 4095)
+			Cvar_SetValue ("gl_fogstart", 4095);
+		if (gl_fogend.value < 6)
+			Cvar_SetValue ("gl_fogend", 6);
+		if (gl_fogend.value > 4096)
+			Cvar_SetValue ("gl_fogend", 4096);
+		if (gl_fogstart.value >= gl_fogend.value)
+			Cvar_SetValue ("gl_fogend", gl_fogstart.value + 1);
+
+		fog_colours[0] = (GLfloat) gl_fogred.value;
+		fog_colours[1] = (GLfloat) gl_foggreen.value;
+		fog_colours[2] = (GLfloat) gl_fogblue.value;
+		fog_colours[3] = (GLfloat) 1;//gl_fogalpha.value;
+
+		glFogi_fp (GL_FOG_MODE, GL_LINEAR);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP);
+	//	glFogi_fp (GL_FOG_MODE, GL_EXP2);
+		glFogfv_fp (GL_FOG_COLOR, fog_colours);
+		glFogf_fp (GL_FOG_START, gl_fogstart.value);
+		glFogf_fp (GL_FOG_END, gl_fogend.value);
+		glFogf_fp (GL_FOG_DENSITY, gl_fogdensity.value);
+	//	glHint_fp (GL_FOG_HINT, GL_NICEST);
+		glEnable_fp (GL_FOG);
+	}
+	else
+	{
+		glDisable_fp (GL_FOG);
+	}
+
 	// render normal view
 	R_RenderScene ();
 
@@ -1890,6 +1957,9 @@ void R_RenderView (void)
 
 	R_DrawViewModel();
 
+	if (gl_fogenable.value)
+		glDisable_fp(GL_FOG);
+
 	// render mirror view
 //	R_Mirror ();
 
@@ -1899,6 +1969,8 @@ void R_RenderView (void)
 	{
 		R_PrintTimes ();
 	}
+
+	glDepthMask_fp(1);
 }
 
 void R_DrawName(vec3_t origin, char *Name, int Red)
diff -urNp uhexen2-20060112/hexenworld/Client/gl_rmisc.c uhexen2-20060112b/hexenworld/Client/gl_rmisc.c
--- uhexen2-20060112/hexenworld/Client/gl_rmisc.c	2005-10-25 23:04:17.000000000 +0300
+++ uhexen2-20060112b/hexenworld/Client/gl_rmisc.c	2006-01-12 15:59:53.000000000 +0200
@@ -196,6 +196,15 @@ void R_Init (void)
 	Cvar_RegisterVariable (&gl_multitexture);
 	Cvar_RegisterVariable (&r_teamcolor);
 
+	Cvar_RegisterVariable (&gl_fogenable);
+	Cvar_RegisterVariable (&gl_fogstart);
+	Cvar_RegisterVariable (&gl_fogend);
+	Cvar_RegisterVariable (&gl_fogdensity);
+	Cvar_RegisterVariable (&gl_fogalpha);
+	Cvar_RegisterVariable (&gl_fogred);
+	Cvar_RegisterVariable (&gl_fogblue);
+	Cvar_RegisterVariable (&gl_foggreen);
+
 	Cvar_RegisterVariable (&gl_glows);
 	Cvar_RegisterVariable (&gl_missile_glows);
 	Cvar_RegisterVariable (&gl_other_glows);
